File: HeirloomsWeb/src/crypto/vaultSession.js
const MEM_MAX = 300

let _masterKey = null
let _sharingPrivkey = null
const _plotKeys = new Map() // plotId → raw 32-byte Uint8Array

export const thumbnailCache = new Map()

export function unlock(masterKey) {
  _masterKey = masterKey
}

export function setSharingPrivkey(cryptoKey) {
  _sharingPrivkey = cryptoKey
}

export function getSharingPrivkey() {
  return _sharingPrivkey
}

export function setPlotKey(plotId, rawKeyBytes) {
  _plotKeys.set(plotId, rawKeyBytes)
}

export function getPlotKey(plotId) {
  return _plotKeys.get(plotId) ?? null
}

export function lock() {
  _masterKey = null
  _sharingPrivkey = null
  _plotKeys.clear()
  thumbnailCache.forEach(url => URL.revokeObjectURL(url))
  thumbnailCache.clear()
}

export function isUnlocked() {
  return _masterKey !== null
}

export function getMasterKey() {
  if (!_masterKey) throw new Error('Vault is locked')
  return _masterKey
}

export function cacheThumbnail(uploadId, objectUrl) {
  if (thumbnailCache.size >= MEM_MAX) {
    const oldest = thumbnailCache.keys().next().value
    URL.revokeObjectURL(thumbnailCache.get(oldest))
    thumbnailCache.delete(oldest)
  }
  thumbnailCache.set(uploadId, objectUrl)
}

File: HeirloomsApp/app/src/main/kotlin/digital/heirlooms/crypto/VaultSession.kt
package digital.heirlooms.crypto

import androidx.compose.ui.graphics.ImageBitmap
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import java.util.concurrent.ConcurrentHashMap

object VaultSession {

    @Volatile private var _masterKey: ByteArray? = null
    @Volatile private var _sharingPrivkey: ByteArray? = null

    val isUnlocked: Boolean get() = _masterKey != null

    // Throws if vault is not unlocked.
    val masterKey: ByteArray
        get() = _masterKey?.copyOf() ?: error("Vault is not unlocked")

    // PKCS8 DER bytes of the account-level sharing private key.
    // Null until loaded from server after vault unlock.
    val sharingPrivkey: ByteArray?
        get() = _sharingPrivkey?.copyOf()

    // In-memory cache of decrypted thumbnails, keyed by upload ID.
    // Cleared when the vault is locked (process restart).
    val thumbnailCache: ConcurrentHashMap<String, ImageBitmap> = ConcurrentHashMap()

    // Per-plot raw AES-256-GCM plot keys, keyed by plot ID.
    // Populated lazily on first shared-plot access. Cleared on vault lock.
    val plotKeys: ConcurrentHashMap<String, ByteArray> = ConcurrentHashMap()

    // Increments each time a plot key is added. Observed by HeirloomsImage to retry
    // thumbnail decryption for items whose key wasn't cached on first render.
    private val _plotKeysVersion = MutableStateFlow(0)
    val plotKeysVersion: StateFlow<Int> get() = _plotKeysVersion

    fun setPlotKey(plotId: String, rawKeyBytes: ByteArray) {
        plotKeys[plotId] = rawKeyBytes.copyOf()
        _plotKeysVersion.value++
    }

    fun getPlotKey(plotId: String): ByteArray? = plotKeys[plotId]?.copyOf()

    fun unlock(masterKey: ByteArray) {
        _masterKey = masterKey.copyOf()
    }

    fun setSharingPrivkey(pkcs8Bytes: ByteArray) {
        _sharingPrivkey = pkcs8Bytes.copyOf()
    }

    fun lock() {
        _masterKey?.fill(0)
        _masterKey = null
        _sharingPrivkey?.fill(0)
        _sharingPrivkey = null
        thumbnailCache.clear()
        plotKeys.values.forEach { it.fill(0) }
        plotKeys.clear()
        _plotKeysVersion.value = 0
    }
}
